collector/exchange.go

```diff
@@ -0,0 +1,679 @@
+		"database_instances",
+		"http_proxy",
+		"activesync",
+		"availability_service",
+		"owa",
+		"autodiscover",
+		"management_workloads",
+		"rpc",
+	}
+
+	exchangeCollectorFuncDesc map[string]string = map[string]string{
+		"ldap":                 "(WMI Class: win32_PerfRawData_MSExchangeADAccess_MSExchangeADAccessProcesses)",
+		"transport_queues":     "(WMI Class: win32_PerfRawData_MSExchangeTransportQueues_MSExchangeTransportQueues)",
+		"database_instances":   "(WMI Class: win32_PerfRawData_ESE_MSExchangeDatabaseInstances)",
+		"http_proxy":           "(WMI Class: win32_PerfRawData_MSExchangeHttpProxy_MSExchangeHttpProxy)",
+		"activesync":           "(WMI Class: win32_PerfRawData_MSExchangeActiveSync_MSExchangeActiveSync)",
+		"availability_service": "(WMI Class: win32_PerfRawData_MSExchangeAvailabilityService_MSExchangeAvailabilityService)",
+		"owa":                  "(WMI Class: win32_PerfRawData_MSExchangeOWA_MSExchangeOWA)",
+		"autodiscover":         "(WMI Class: win32_PerfRawData_MSExchangeAutodiscover_MSExchangeAutodiscover)",
+		"management_workloads": "(WMI Class: win32_PerfRawData_MSExchangeWorkloadManagementWorkloads_MSExchangeWorkloadManagementWorkloads)",
+		"rpc":                  "(WMI Class: win32_PerfRawData_MSExchangeRpcClientAccess_MSExchangeRpcClientAccess)",
+	}
+
+	argExchangeListAllCollectors = kingpin.Flag(
+		"collectors.exchange.list",
+		"Lists all available exchange collectors and their description",
+	).Bool()
+
+	argExchangeEnabledCollectors = kingpin.Flag(
+		"collectors.exchange.enable",
+		"comma-separated list of exchange collectors to use",
+	).Default(strings.Join(exchangeAllCollectorFuncs, ",")).String()
+
+	argExchangeDisabledCollectors = kingpin.Flag(
+		"collectors.exchange.disable",
+		"comma-separated list of exchange collectors NOT to use",
+	).Default().String()
+)
+
+func init() {
+	registerCollector("exchange", newExchangeCollector)
+}
+
+// desc creates a new prometheus description
+func desc(metricName string, description string, labels ...string) *prometheus.Desc {
+	return prometheus.NewDesc(
+		prometheus.BuildFQName(Namespace, subsystem, metricName),
+		description,
+		labels,
+		nil,
+	)
+}
+
+// contains checks if element e exists in slice s
+func contains(s []string, e string) bool {
+	for _, a := range s {
+		if a == e {
+			return true
+		}
+	}
+	return false
+}
+
+// toLabelName converts strings to lowercase and replaces all whitespace and dots with underscores
+func toLabelName(name string) string {
+	return strings.ReplaceAll(strings.Join(strings.Fields(strings.ToLower(name)), "_"), ".", "_")
+}
+
+// newExchangeCollector returns a new Collector
+func newExchangeCollector() (Collector, error) {
+	// https://docs.microsoft.com/en-us/exchange/exchange-2013-performance-counters-exchange-2013-help
+	c := exchangeCollector{
+		// MS Exchange RPC Client Access
+		RPCAveragedLatency:  desc("rpc_avg_latency", "The latency (ms), averaged for the past 1024 packets"),
```

Comment:
Two points here:
1. Is the reported value actually an average? There's frequently confusion around names which do not reflect the raw data, and this could be one of them.
2. Make to have the unit in the name, and also to use base units. Ie seconds in this case, so you'll need to divide the value read by 1000.

Reply:
1. [According to Microsoft](https://docs.microsoft.com/en-us/exchange/exchange-2013-performance-counters-exchange-2013-help) it's the latency averaged for the past 1024 packets. I haven't done any further investigations

2. I'll update the unit names/values where applicable

Reply:
Yes, but Microsoft only documents how it is intended to be visualised, not the actual raw data. 
There's some docs [here](https://docs.microsoft.com/en-us/previous-versions/windows/it-pro/windows-server-2003/cc785636(v=ws.10)) on the data types. In previous cases of similarly-named fields, we've seen this as a sum-and-count pair, so you get the total number of milliseconds in this case, and how many operations those are measured over.